lib/src/repo.rs

```diff
@@ -241,6 +242,40 @@ impl ReadonlyRepo {
         })
     }
 
+    // An interface for testing this functionality directly is constructed in
+    // a follow-up commit.
+    pub fn shortest_unique_prefix_length(&self, target_id_hex: &str) -> usize {
```

Comment:
Can we put this on the index instead? Then it will work for `MutableRepo`s too. It may not usually matter, but I also don't think it should be hard to do. Putting it on the index also means that we can take advantage of the lookup table it already has. I think this is the sort of thing the index should be responsible for, so different index implementations can do it differently. In particular, we'll clearly not want to iterate over all commits in Google's internal repo :)

Reply:
I certainly agree it belongs to the index. Is there a way which change ids are visible from inside the index? I would need something to replace

```rust
       let all_visible_revisions = RevsetExpression::all()
            .evaluate(self.as_repo_ref(), None)
            .unwrap();
        let change_hex_iter = all_visible_revisions
            .iter()
            .map(|index_entry| index_entry.change_id().hex());
```

I don't know if there's an easy way to give an `IndexEntry` the ability to know whether it's visible. 

We could also consider changing the behavior of ignoring invisible change ids, but that's a larger change, and I'm not sure if it's for the better.

Reply:
Ah, I forgot about change IDs. I don't remember what we said last time, but do you think it makes sense to have one method on the index for finding a unique prefix among all indexed commit IDs, and another method for a unique prefix among all change IDs reachable from a given set of heads? Then the caller can pass in the visible heads to that second method. I feel like it's better to keep the searches among commit IDs and change IDs separate, so the caller can decide. For example, we might add `commit_id(ab12)` and `change_id(ab12)` revsets one day, and they shouldn't have to resolve the prefixes among the other kind of ID. What do you think?

Reply:
From the algorithmic point of view, the important thing is that the change ids should be in an index, and in the end state it should be efficiently stored on disk. Currently, all the trie/BTree stuff is mainly needed because change ids aren't stored in an optimized (sorted, at least) fashion on disk.

It would make sense the repo to only store an in-memory index of change ids and use the other index for commit ids. I didn't do it because it makes the logic slightly harder than when everything is in the same bucket, and there's nothing in `jj` that would use the separation just yet.

This is not strictly relevant to this PR, but I think I was also leaning towards putting everything in the same bucket after you told me that the lookup by prefix in the Index is inefficient. It was simpler for me to plan to avoid it than to figure out how to fix it. It would just take adding a few lines to `IdIndex` using either a BTree or a trie.

From the UI point of view, I was thinking of the "IdIndex" I was building as more of a precursor to  what we were talking about https://github.com/martinvonz/jj/pull/994#discussion_r1070874338: a unified place responsible for the logic of how user-specified prefixes correspond to the thing they want to find (usually a change or an id). Such a thing should exist, whether it stores the ids itself or communicates with two different indices is less important. I'm itching to make the same entity responsible for the shortest prefix computation (not the algorithmic part, but the part that controls the logic) and for looking up a commit/change by prefix.

In other words, while eventually different entities should probably be responsible for the algorithmic part and the UI part, for now I tried to make one thing that will do both.

Reply:
That's fair. We're going to have to rewrite the index soon and then we'll revisit this. Until then, it sounds good to me to just do the simplest reasonable thing.